### PR TITLE
fix: flowchart rendering issues

### PR DESCRIPTION
**Pull request by igorwessel**
Closes https://github.com/excalidraw/mermaid-to-excalidraw/issues/45

# Revision

## 8d58faa38229d2b30b2301f1173c85f525d1995b

There are cases where one entity can have multiple relationships to another entity, we were only considering that there was just one relationship.

## 31fea663c573940876425ed04e95732955d3b4fb

The filter condition was updated to include points that change in either the x or y direction. 
This ensures that points on a straight line in the same direction are included in the reflectionPoints array, and ensures line commands when only changes direction.

By ignoring the directions that a line command can make we were only having straight lines. 

| Before | After |
|------|-----|
| Captura de Tela 2024-05-06 às 13 48 26 | Captura de Tela 2024-05-06 às 13 48 48|

# Testcases

Captura de Tela 2024-05-06 às 07 18 48

### CODE REVIEW

**Review by ad1992 — CHANGES_REQUESTED** [review]
@igorwessel lets add a testcase for the same in playground

File: src/parser/flowchart.ts [hunk 1]
```diff
@@ -228,9 +240,12 @@ export const parseMermaidFlowChartDiagram = (
   Object.keys(vertices).forEach((id) => {
     vertices[id] = parseVertex(vertices[id], containerEl);
   });
+
+  const processedEdges = new Set<string>();
```
Comment (igorwessel) [inline src/parser/flowchart.ts:244]: **thought:** ‏I feel there's a better way to do this but I haven't figured it out yet, the same happens in the state diagram.

Comment (ad1992) [reply src/parser/flowchart.ts]: I have made into a function closure for better encapsulation

**Review by ad1992 — CHANGES_REQUESTED** [review]
Thanks for the fix @igorwessel 
I have turned processedEdges into a function closure for better encapsulation

The arrows seems to be breaking in this PR 👇🏻 

image

**Comment by ad1992** [comment]
> @ad1992 It was also mentioned about escaping HTML. Does it make sense for mermaid-to-excalidraw to bring this as well? I see that this is a bit out of the scope of this PR, it has to be something a bit more global for all diagrams.

I guess you meant font awesome support ? we don't support it yet, it's out of scope of this PR. We can look into that later.

**Comment by igorwessel** [comment]
>I guess you meant font awesome support ? we don't support it yet, it's out of scope of this PR. We can look into that later.

No but basic support for some HTML tags, for example: `<br>`, although you could use `\n`. I don't know to what extent we have possibilities here.

In this issue #45, we have some entities using HTML tags to format labels, although I believe we should not support this.
Captura de Tela 2024-05-07 às 07 03 14

**Comment by igorwessel** [comment]
>The arrows seems to be breaking in this PR 👇🏻

@ad1992 Apparently, the problem is that the penultimate point is very close to the last point. It may be solved by calculating the distance between the penultimate point and the last point, and only including the penultimate point when it is greater than some limit we define. 
But I feel there might be something more that I'm not seeing at the moment.

Captura de Tela 2024-05-07 às 07 46 06

**Comment by ad1992** [comment]
> > The arrows seems to be breaking in this PR 👇🏻
> 
> @ad1992 Apparently, the problem is that the penultimate point is very close to the last point. It may be solved by calculating the distance between the penultimate point and the last point, and only including the penultimate point when it is greater than some limit we define. But I feel there might be something more that I'm not seeing at the moment.
> 
> Captura de Tela 2024-05-07 às 07 46 06

@igorwessel this was working earlier for all arrows but not anymore. Since it is breaking post the changes in this PR, we should be able to fix it in this PR itself.

**Comment by igorwessel** [comment]
@ad1992 I implemented two solutions but I'm still undecided about which is the best approach:

1. In the penultimate point, we check if the previous point in some axis remains the same, if it remains the same we can affirm that it is a straight line either on the X or Y axis. In this case we can just return as false.
2. We check the penultimate point but we set a minimum distance for it to be included, maybe it makes sense to do this for all the points that we identify as a straight line?

I sent the commit with both solutions, both have their pros/cons but should work in the majority of cases.

And I should check the closure because apparently the problem of repeated edges has returned.

**Comment by igorwessel** [comment]
I'm will pull the updates from main, and create some tests to ensure.

File: src/parser/flowchart.ts (hunk 1 above)
Comment (igorwessel) [reply src/parser/flowchart.ts]: I added a map with the aim of avoiding a for loop inside parseEdges, and the only way I thought to encapsulate it was to create a kind of factory haha.
```js
function createEdgeParser() {
  const edgeCountMap = new Map<string, number>();

  return function parseEdge(data: any, containerEl: HTMLElement) {
    const edgeIdentifier = `${data.start}-${data.end}`;
    const count = edgeCountMap.get(edgeIdentifier) || 0;
    edgeCountMap.set(edgeIdentifier, count + 1);

    // Rest of parseEdge implementation...

    return {

    };
  };
}

**Comment by igorwessel** [comment]
@ad1992 friendly reminder, you think this distance calculation solution is enough?

**Comment by ad1992** [comment]
> @ad1992 friendly reminder, you think this distance calculation solution is enough?

Hi @igorwessel sorry for the delayed response! I am gonna check this soon either tonight or tomorrow morning

**Review by ad1992 — COMMENTED** [review]
@igorwessel looks good!
I have pushed some tweaks and improvements in https://github.com/excalidraw/mermaid-to-excalidraw/pull/61/commits/e7e9f65e95c1f93709ca092fd1e8856ce66d2e92, let me know if this looks good and we can update the tests then accordingly.

**Comment by igorwessel** [comment]
@ad1992 it's great! I didn't remember there was a function for hypotenuse in Math haha

**Comment by ad1992** [comment]
Pushed a fix for excluding point when its same as prev point
Merging, thanks @igorwessel ✨

**Review by ad1992 — APPROVED** [review]
🚀

**Pull request merged**